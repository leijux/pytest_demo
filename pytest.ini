[pytest]
testpaths = testcases/
markers =
    single: single api test
    multiple: multiple api test
    negative: abnormal test case

;xfail_strict = True
;--snapshot-update --clean-alluredir
addopts = --alluredir reports/allure_results --snapshot-update --allow-snapshot-deletion --clean-alluredir

log_auto_indent = True

log_cli = True
log_cli_level = info
log_cli_format = %(asctime)s %(filename)s:%(lineno)s [%(levelname)s]: %(message)s
log_cli_date_format = %Y-%m-%d %H:%M:%S

log_file = logs/test-{}.log
log_file_level = debug
log_file_format = %(asctime)s %(filename)s:%(lineno)s [%(levelname)s]: %(message)s
log_file_date_format = %Y-%m-%d %H:%M:%S

;function module
asyncio_default_fixture_loop_scope = module