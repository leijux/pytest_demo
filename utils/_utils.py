import json
import time


def json_dumps(json_data: dict) -> str:
    # Python3中，json在做dumps操作时，会将中文转换成unicode编码，因此设置 ensure_ascii=False
    return json.dumps(json_data, indent=4, ensure_ascii=False, sort_keys=True)


def timestamp() -> int:
    return int(round(time.time() * 1000))


def dict_to_csv(d: dict) -> str:
    return f"{'\n'.join([f"{key},\"{value}\"," for key, value in d.items()])}"
