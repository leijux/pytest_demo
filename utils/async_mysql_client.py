import aiomysql
from typing import List, Dict, Any, Optional, AsyncGenerator


class MySQLClient:
    """异步MySQL数据库操作工具类"""

    def __init__(
            self,
            host: str,
            port: int,
            database: str,
            user: str,
            password: str,
            **kwargs
    ):
        """
        初始化数据库连接参数
        :param host: 数据库地址
        :param port: 端口号
        :param database: 数据库名称
        :param user: 用户名
        :param password: 密码
        :param kwargs: 其他连接参数
        """
        self.host = host
        self.port = port
        self.database = database
        self.user = user
        self.password = password
        self.conn_args = kwargs
        self.pool: Optional[aiomysql.Pool] = None

    async def __aenter__(self):
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def connect(self):
        """创建连接池（推荐使用连接池）"""
        try:
            self.pool = await aiomysql.create_pool(
                host=self.host,
                port=self.port,
                db=self.database,
                user=self.user,
                password=self.password,
                autocommit=False,
                **self.conn_args
            )
        except aiomysql.OperationalError as e:
            raise ConnectionError(f"数据库连接失败: {str(e)}")

    async def close(self):
        """关闭所有连接"""
        if self.pool:
            self.pool.close()
            await self.pool.wait_closed()

    async def execute_query(
            self,
            query: str,
            params: Optional[dict] = None
    ) -> List[Dict[str, Any]]:
        """
        执行查询语句（异步）
        :param query: SQL查询语句
        :param params: 查询参数（字典格式）
        :return: 查询结果列表（字典形式）
        """
        if not self.pool:
            raise RuntimeError("数据库未连接")

        async with self.pool.acquire() as conn:
            async with conn.cursor(aiomysql.DictCursor) as cursor:
                try:
                    await cursor.execute(query, params)
                    return await cursor.fetchall()
                except aiomysql.Error as e:
                    await conn.rollback()
                    raise RuntimeError(f"查询执行失败: {str(e)}")

    async def execute_update(
            self,
            query: str,
            params: Optional[dict] = None
    ) -> int:
        """
        执行更新操作（异步）
        :param query: SQL语句
        :param params: 参数（字典格式）
        :return: 受影响的行数
        """
        if not self.pool:
            raise RuntimeError("数据库未连接")

        async with self.pool.acquire() as conn:
            async with conn.cursor() as cursor:
                try:
                    await cursor.execute(query, params)
                    await conn.commit()
                    return cursor.rowcount
                except aiomysql.Error as e:
                    await conn.rollback()
                    raise RuntimeError(f"更新操作失败: {str(e)}")

    async def batch_insert(
            self,
            table: str,
            data: List[Dict[str, Any]]
    ) -> int:
        """
        批量插入数据（异步）
        :param table: 表名
        :param data: 数据字典列表
        :return: 插入的行数
        """
        if not data:
            return 0

        columns = list(data[0].keys())
        values = [tuple(item.values()) for item in data]
        placeholders = ','.join(['%s'] * len(columns))

        query = f"INSERT INTO `{table}` ({', '.join(columns)}) VALUES ({placeholders})"

        async with self.pool.acquire() as conn:
            async with conn.cursor() as cursor:
                try:
                    await cursor.executemany(query, values)
                    await conn.commit()
                    return cursor.rowcount
                except aiomysql.Error as e:
                    await conn.rollback()
                    raise RuntimeError(f"批量插入失败: {str(e)}")

    async def table_exists(self, table_name: str) -> bool:
        """检查表是否存在（异步）"""
        query = """
            SELECT COUNT(*) as count 
            FROM information_schema.tables
            WHERE table_schema = DATABASE()
            AND table_name = %(table_name)s
        """
        result = await self.execute_query(query, {"table_name": table_name})
        return result[0]['count'] > 0

    # 添加事务管理
    async def transaction(self, queries: List[tuple]):
        """执行事务操作"""
        async with self.pool.acquire() as conn:
            async with conn.cursor() as cursor:
                try:
                    await conn.begin()
                    for query, params in queries:
                        await cursor.execute(query, params)
                    await conn.commit()
                except aiomysql.Error:
                    await conn.rollback()
                    raise

    # 添加流式查询
    async def stream_query(self, query: str, params: dict = None):
        """流式读取大量数据"""
        async with self.pool.acquire() as conn:
            async with conn.cursor(aiomysql.SSCursor) as cursor:
                await cursor.execute(query, params)
                while True:
                    rows = await cursor.fetchmany(1000)
                    if not rows:
                        break
                    for row in rows:
                        yield row
# pytest测试示例 ##################################################

# conftest.py配置
# @pytest.fixture
# async def mysql_client() -> AsyncGenerator[MySQLClient, None]:
#     client = MySQLClient(
#         host="localhost",
#         port=3306,
#         database="testdb",
#         user="root",
#         password="your_password",
#         pool_recycle=3600,
#         pool_size=5
#     )
#     await client.connect()
#     yield client
#     await client.close()
#
#
# # 测试用例（需要pytest-asyncio插件）
# @pytest.mark.asyncio
# async def test_mysql_operations(mysql_client):
#     # 测试表存在检查
#     assert await mysql_client.table_exists("users") is True
#
#     # 测试数据插入
#     test_data = {"username": "test_user", "email": "test@example.com"}
#     insert_sql = """
#         INSERT INTO users (username, email)
#         VALUES (%(username)s, %(email)s)
#     """
#     affected = await mysql_client.execute_update(insert_sql, test_data)
#     assert affected == 1
#
#     # 验证数据
#     select_sql = "SELECT * FROM users WHERE username = %(username)s"
#     results = await mysql_client.execute_query(select_sql, {"username": "test_user"})
#     assert len(results) == 1
#     assert results[0]["email"] == "test@example.com"
#
#     # 清理测试数据
#     delete_sql = "DELETE FROM users WHERE username = %(username)s"
#     await mysql_client.execute_update(delete_sql, {"username": "test_user"})
