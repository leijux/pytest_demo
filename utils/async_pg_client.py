import psycopg
from psycopg import sql
from psycopg.abc import Query
from psycopg.rows import dict_row
from typing import List, Dict, Any, Optional, AsyncGenerator
from typing import LiteralString


class PostgreSQLClient:
    """异步PostgreSQL操作工具类（psycopg3实现）"""

    def __init__(
            self,
            host: str,
            port: int,
            database: str,
            user: str,
            password: str
    ):
        """
        初始化数据库连接参数
        :param host: 数据库地址
        :param port: 端口号
        :param database: 数据库名称
        :param user: 用户名
        :param password: 密码
        """
        self.dsn = f"dbname={database} user={user} password={password} host={host} port={port}"
        self.conn: Optional[psycopg.AsyncConnection] = None

    async def __aenter__(self):
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def connect(self):
        """建立异步数据库连接"""
        try:
            self.conn = await psycopg.AsyncConnection.connect(
                self.dsn,
                row_factory=dict_row  # 返回字典格式结果
            )
        except psycopg.OperationalError as e:
            raise ConnectionError(f"数据库连接失败: {str(e)}")

    async def close(self):
        """关闭数据库连接"""
        if self.conn:
            await self.conn.close()

    async def execute_query(
            self,
            query: Query,
            params: Optional[dict] = None
    ) -> List[Dict[str, Any]]:
        """
        执行查询语句
        :param query: SQL查询语句
        :param params: 查询参数（字典格式）
        :return: 查询结果列表（字典形式）
        """
        if not self.conn:
            raise RuntimeError("数据库未连接")

        async with self.conn.cursor() as cursor:
            try:
                await cursor.execute(query, params)
                return await cursor.fetchall()
            except psycopg.Error as e:
                await self.conn.rollback()
                raise RuntimeError(f"查询执行失败: {str(e)}")

    async def execute_update(
            self,
            query: Query,
            params: Optional[dict] = None
    ) -> int:
        """
        执行更新操作
        :param query: SQL语句
        :param params: 参数（字典格式）
        :return: 受影响的行数
        """
        if not self.conn:
            raise RuntimeError("数据库未连接")

        async with self.conn.cursor() as cursor:
            try:
                await cursor.execute(query, params)
                await self.conn.commit()
                return cursor.rowcount
            except psycopg.Error as e:
                await self.conn.rollback()
                raise RuntimeError(f"更新操作失败: {str(e)}")

    async def batch_insert(
            self,
            table: str,
            data: List[Dict[str, Any]]
    ) -> int:
        """
        批量插入数据
        :param table: 表名
        :param data: 数据字典列表
        :return: 插入的行数
        """
        if not data:
            return 0

        columns = list(data[0].keys())
        values = [tuple(item.values()) for item in data]

        query = sql.SQL("INSERT INTO {} ({}) VALUES ({})").format(
            sql.Identifier(table),
            sql.SQL(', ').join(map(sql.Identifier, columns)),
            sql.SQL(', ').join([sql.Placeholder()] * len(columns))
        )

        if not self.conn:
            raise RuntimeError("数据库未连接")

        try:
            async with self.conn.cursor() as cursor:
                await cursor.executemany(query, values)
                await self.conn.commit()
                return cursor.rowcount
        except psycopg.Error as e:
            await self.conn.rollback()
            raise RuntimeError(f"批量插入失败: {str(e)}")

    async def table_exists(self, table_name: str) -> bool:
        """检查表是否存在"""
        query: LiteralString = """
            SELECT EXISTS (
                SELECT FROM pg_tables
                WHERE schemaname = 'public'
                AND tablename = %(table_name)s
            )
        """

        result = await self.execute_query(query, {"table_name": table_name})
        return result[0]['exists']

# pytest异步测试示例 ##############################################

# # conftest.py配置
# @pytest.fixture
# async def async_db() -> AsyncGenerator[AsyncPostgreSQL, None]:
#     db = AsyncPostgreSQL(
#         host="localhost",
#         port=5432,
#         database="testdb",
#         user="postgres",
#         password="your_password"
#     )
#     await db.connect()
#     yield db
#     await db.close()
#
#
# # 测试用例示例（需要pytest-asyncio插件）
# @pytest.mark.asyncio
# async def test_async_query(async_db):
#     # 测试表存在检查
#     assert await async_db.table_exists("users") is True
#
#     # 测试插入数据
#     test_data = {"username": "test_user", "email": "test@example.com"}
#     insert_sql = """
#         INSERT INTO users (username, email)
#         VALUES (%(username)s, %(email)s)
#     """
#     affected = await async_db.execute_update(insert_sql, test_data)
#     assert affected == 1
#
#     # 验证插入结果
#     query_sql = "SELECT * FROM users WHERE username = %(username)s"
#     results = await async_db.execute_query(query_sql, {"username": "test_user"})
#     assert len(results) == 1
#     assert results[0]["email"] == "test@example.com"
#
#     # 清理测试数据
#     delete_sql = "DELETE FROM users WHERE username = %(username)s"
#     await async_db.execute_update(delete_sql, {"username": "test_user"})
