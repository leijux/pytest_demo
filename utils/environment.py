import os


class Environment:
    def __init__(self):
        self._base_url: str | None = os.getenv("BASE_URL")

        for key, value in self.__dict__.items():
            if value is None:
                raise ValueError(f"Environment variable {key} is empty")

    @property
    def BASE_URL(self) -> str:
        return self._base_url


env = Environment()
