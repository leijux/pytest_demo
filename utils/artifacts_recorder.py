import asyncio
import allure
import hashlib
import shutil
import os
from pathlib import Path
from typing import (
    Any,
    List,
)

import pytest
from playwright.async_api import (
    BrowserContext,
    Error,
    Page,
    Playwright,
)
from slugify import slugify
import tempfile
import pytest_playwright_asyncio


class ArtifactsRecorder:
    def __init__(
        self,
        pytestconfig: Any,
        request: pytest.FixtureRequest,
        output_path: str,
        playwright: Playwright,
        pw_artifacts_folder: tempfile.TemporaryDirectory,
    ) -> None:
        self._request = request
        self._pytestconfig = pytestconfig
        self._playwright = playwright
        self._output_path = output_path
        self._pw_artifacts_folder = pw_artifacts_folder

        self._all_pages: List[Page] = []
        self._screenshots: List[str] = []
        self._traces: List[str] = []
        self._tracing_option = pytestconfig.getoption("--tracing")
        self._capture_trace = self._tracing_option in [
            "on", "retain-on-failure"]

    def _build_artifact_test_folder(self, folder_or_file_name: str) -> str:
        return os.path.join(
            self._output_path,
            _truncate_file_name(folder_or_file_name),
        )

    async def did_finish_test(self, failed: bool) -> None:
        screenshot_option = self._pytestconfig.getoption("--screenshot")
        capture_screenshot = screenshot_option == "on" or (
            failed and screenshot_option == "only-on-failure"
        )
        if capture_screenshot:
            for index, screenshot in enumerate(self._screenshots):
                human_readable_status = "failed" if failed else "finished"
                screenshot_name = f"test-{human_readable_status}-{index + 1}.png"
                screenshot_path = self._build_artifact_test_folder(
                    screenshot_name,
                )
                # os.makedirs(os.path.dirname(screenshot_path), exist_ok=True)
                # shutil.move(screenshot, screenshot_path)
                allure.attach.file(
                    screenshot_path,
                    name=screenshot_name,
                    attachment_type=allure.attachment_type.PNG,
                )
        else:
            for screenshot in self._screenshots:
                os.remove(screenshot)

        if self._tracing_option == "on" or (
            failed and self._tracing_option == "retain-on-failure"
        ):
            for index, trace in enumerate(self._traces):
                trace_file_name = (
                    "trace.zip" if len(
                        self._traces) == 1 else f"trace-{index + 1}.zip"
                )
                trace_path = self._build_artifact_test_folder(trace_file_name)
                os.makedirs(os.path.dirname(trace_path), exist_ok=True)
                shutil.move(trace, trace_path)
        else:
            for trace in self._traces:
                os.remove(trace)

        video_option = self._pytestconfig.getoption("--video")
        preserve_video = video_option == "on" or (
            failed and video_option == "retain-on-failure"
        )
        if preserve_video:
            for index, page in enumerate(self._all_pages):
                video = page.video
                if not video:
                    continue
                try:
                    video_file_name = (
                        "video.webm"
                        if len(self._all_pages) == 1
                        else f"video-{index + 1}.webm"
                    )
                    path = self._build_artifact_test_folder(video_file_name)
                    await video.save_as(
                        path=path
                    )
                    # with open(path, "rb") as f:
                    allure.attach(path, name=video_file_name,
                                  attachment_type=allure.attachment_type.WEBM)
                    await asyncio.sleep(5)
                    await page.video.delete()
                except Error as e:
                    # Silent catch empty videos.

                    pytest.fail(f"Error while saving video: {e}")
        else:
            for page in self._all_pages:
                # Can be changed to "if page.video" without try/except once https://github.com/microsoft/playwright-python/pull/2410 is released and widely adopted.
                if video_option in ["on", "retain-on-failure"]:
                    try:
                        if page.video:
                            await page.video.delete()
                    except Error:
                        pass

    async def on_did_create_browser_context(self, context: BrowserContext) -> None:
        context.on("page", lambda page: self._all_pages.append(page))
        if self._request and self._capture_trace:
            await context.tracing.start(
                title=slugify(self._request.node.nodeid),
                screenshots=True,
                snapshots=True,
                sources=True,
            )

    async def on_will_close_browser_context(self, context: BrowserContext) -> None:
        if self._capture_trace:
            trace_path = Path(self._pw_artifacts_folder.name) / _create_guid()
            await context.tracing.stop(path=trace_path)
            self._traces.append(str(trace_path))
        else:
            await context.tracing.stop()

        if self._pytestconfig.getoption("--screenshot") in ["on", "only-on-failure"]:
            for page in context.pages:
                try:
                    screenshot_path = (
                        Path(self._pw_artifacts_folder.name) / _create_guid()
                    )
                    screenshot = await page.screenshot(
                        timeout=5000,
                        full_page=self._pytestconfig.getoption(
                            "--full-page-screenshot"
                        ),
                    )
                    allure.attach(screenshot, name=screenshot_path.stem,
                                  attachment_type=allure.attachment_type.PNG)

                    self._screenshots.append(str(screenshot_path))
                except Error:
                    pass


pytest_playwright_asyncio.ArtifactsRecorder = ArtifactsRecorder


def _create_guid() -> str:
    return hashlib.sha256(os.urandom(16)).hexdigest()


def _truncate_file_name(file_name: str) -> str:
    if len(file_name) < 256:
        return file_name
    return f"{file_name[:100]}-{hashlib.sha256(file_name.encode()).hexdigest()[:7]}-{file_name[-100:]}"
