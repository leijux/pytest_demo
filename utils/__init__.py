from utils.read_data import DataCache, DataTest, test_data
from utils.environment import env
from utils._utils import timestamp, json_dumps, dict_to_csv
from utils.step_context import StepContext

rootdir: str
