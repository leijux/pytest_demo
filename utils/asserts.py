import allure
from core import ResultBase
from utils import DataTest
import syrupy


@allure.step("assert")
def assert_result(result: ResultBase, expected: DataTest, snapshot=None):
    """
    执行断言
    :param result: 请求结果
    :param expected: 预期结果
    :param snapshot: 是否使用快照
    :param schema_name: schema名称
    """
    __tracebackhide__ = True
    assert result.response.status_code == expected.except_status_code, f"{
        expected.except_status_code} is not equal {result.response.status_code}"

    assert result.success == expected.except_success, f"{
        expected.except_success} is not equal {result.success}, err: {result.error}"

    if hasattr(expected, "except_msg"):
        assert result.msg == expected.except_msg, f"{
            result.msg} is not equal {expected.except_msg}"

    if hasattr(expected, "except_code"):
        assert result.code == expected.except_code, f" {
            result.code} is not equal {expected.except_code}"

    if snapshot:
        content, content_type = result.content()
        if not content:
            return

        assert content == snapshot
