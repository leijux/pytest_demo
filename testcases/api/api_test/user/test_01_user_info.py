import pytest
import allure
import base64

from utils.asserts import assert_result


@allure.severity(allure.severity_level.NORMAL)
@allure.epic("单接口测试")
@allure.feature("用户模块")
@allure.story("/api/v1/user")
@allure.link("https://gitea.com/api/swagger#/user/userGetCurrent", name="to swagger")
@pytest.mark.asyncio(loop_scope="class")
@pytest.mark.single
class TestUserInfo:

    @allure.title("获取用户信息 {param_id}")
    @allure.description("验证获取当前登录的用户信息接口")
    async def test_user_info(self, user_opn, snapshot, test_data):
        result = await user_opn.get_user_info(base64.b64encode(
            f"{test_data.username}:{test_data.password}".encode()).decode())

        assert_result(result, test_data, snapshot)

    @allure.title("获取用户信息 {param_id}")
    @allure.description("验证获取当前登录的用户信息接口异常basic_auth")
    @pytest.mark.negative
    async def test_user_info_with_basic_auth(self, user_opn, snapshot, test_data):
        basic_auth = test_data.basic_auth or base64.b64encode(
            f"{test_data.username}:{test_data.password}".encode()).decode()

        result = await user_opn.get_user_info(basic_auth)

        assert_result(result, test_data, snapshot)


if __name__ == '__main__':
    pytest.main(["-q", "-s", __file__])
