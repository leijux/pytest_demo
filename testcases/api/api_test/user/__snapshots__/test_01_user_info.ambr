# serializer version: 1
# name: TestUserInfo.test_user_info[正确用户名密码_登录成功]
  '''
  {
      "properties": {
          "active": {
              "type": "boolean"
          },
          "avatar_url": {
              "type": "string"
          },
          "created": {
              "type": "string"
          },
          "description": {
              "type": "string"
          },
          "email": {
              "type": "string"
          },
          "followers_count": {
              "type": "integer"
          },
          "following_count": {
              "type": "integer"
          },
          "full_name": {
              "type": "string"
          },
          "html_url": {
              "type": "string"
          },
          "id": {
              "type": "integer"
          },
          "is_admin": {
              "type": "boolean"
          },
          "language": {
              "type": "string"
          },
          "last_login": {
              "type": "string"
          },
          "location": {
              "type": "string"
          },
          "login": {
              "type": "string"
          },
          "login_name": {
              "type": "string"
          },
          "prohibit_login": {
              "type": "boolean"
          },
          "restricted": {
              "type": "boolean"
          },
          "source_id": {
              "type": "integer"
          },
          "starred_repos_count": {
              "type": "integer"
          },
          "username": {
              "type": "string"
          },
          "visibility": {
              "type": "string"
          },
          "website": {
              "type": "string"
          }
      },
      "required": [
          "active",
          "avatar_url",
          "created",
          "description",
          "email",
          "followers_count",
          "following_count",
          "full_name",
          "html_url",
          "id",
          "is_admin",
          "language",
          "last_login",
          "location",
          "login",
          "login_name",
          "prohibit_login",
          "restricted",
          "source_id",
          "starred_repos_count",
          "username",
          "visibility",
          "website"
      ],
      "type": "object"
  }
  '''
# ---
# name: TestUserInfo.test_user_info_with_basic_auth[密码为空_登录失败]
  '''
  {
      "properties": {
          "message": {
              "type": "string"
          },
          "url": {
              "type": "string"
          }
      },
      "required": [
          "message",
          "url"
      ],
      "type": "object"
  }
  '''
# ---
# name: TestUserInfo.test_user_info_with_basic_auth[用户名为空_登录失败]
  '''
  {
      "properties": {
          "message": {
              "type": "string"
          },
          "url": {
              "type": "string"
          }
      },
      "required": [
          "message",
          "url"
      ],
      "type": "object"
  }
  '''
# ---
# name: TestUserInfo.test_user_info_with_basic_auth[用户名密码均为空_登录失败]
  '''
  {
      "properties": {
          "message": {
              "type": "string"
          },
          "url": {
              "type": "string"
          }
      },
      "required": [
          "message",
          "url"
      ],
      "type": "object"
  }
  '''
# ---
# name: TestUserInfo.test_user_info_with_basic_auth[错误用户名_登录失败]
  '''
  {
      "properties": {
          "message": {
              "type": "string"
          },
          "url": {
              "type": "string"
          }
      },
      "required": [
          "message",
          "url"
      ],
      "type": "object"
  }
  '''
# ---
